### PR TITLE
#56: Add a GHashTable constructor and some helper functions

### PR DESCRIPTION
**Pull request by spk121**
A minimal and primitive layer for GHashTables.... Not sure if it is worth doing more, considering how so much depends on interpreting the pointers that GHashTables use as their keys and values.

### CODE REVIEW

**Comment by LordYuuma** [comment]
Does this fully resolve #56 or would there be more needed?

As far as the Scheme code, that one would write around that, is concerned, it is a little on the verbose side. Given, that you're using the underlying hash table and not converting between SCM Hashes and GHashes, that's understandable, but what is your reason to choose this version over an alternative one? Perhaps performance?

**Comment by spk121** [comment]
Well, it is half the story.
This part of it is to add just enough functionality to deal with hash tables as a type, beyond the methods that GLib provides for GHashTables.
The next part is to decide how to deal with GHashTables as arguments to functions, because GHashTables by themselves don't really have enough information encoded in them for Guile to pack/unpack them without the typelib metainformation.

**Comment by LordYuuma** [comment]
GHashTables may even lack the information to packing/unpacking when considering all the typelib metainfo, given that the HashFunc and EqualFunc (whose behaviour may implicitly be relied on) are lacking from it if I'm not mistaken. In other words, we have to use `<GHashTable>` either way... or do we?

Assuming an implementation on top of `<GHashTable>`, we should be able to at least implement `TRANSFER_NONE` arguments. For `TRANSFER_CONTAINER` and `TRANSFER_EVERYTHING` we will run into memory management issues regardless of what we do.

I think, a preferrable solution to this would be to use SCM hashes and object properties. Specifically, we could store `g-hash-type` as a symbol (`direct`, `string`, `int` ...) or function pointer pair (not yet supported in your code, but if we wanted to have custom hashing functions) and `x-hash-type` as a pair of procedures to use with the `hashx-` functions. If those are not given, we'll just assume `direct` and `eq?` as the weakest ones. That way, we can use good ol' Scheme hashes and have only a little to worry about when converting.

File: src/gig_argument.c [hunk 1]
```diff
@@ -673,6 +678,71 @@ scm_to_c_ptr_array(S2C_ARG_DECL)
...
+        equal_func = g_double_equal;
+    }
+    else if (meta->params[0].gtype == G_TYPE_STRING) {
+        hash_func = g_str_hash;
+        equal_func = g_str_equal;
+    }
+    else {
+        hash_func = NULL;
+        equal_func = NULL;
+    }
```
Comment (LordYuuma) [inline src/gig_argument.c]: That seems to be a good enough guess for now, but we should probably give the user power over this part. (See my rant on object properties).

Comment (spk121) [reply src/gig_argument.c]: I like the idea, but, I just wanted to commit this minimum viable patch before things got more complex.

File: src/gig_argument.c [hunk 2]
```diff
@@ -678,64 +678,184 @@ scm_to_c_ptr_array(S2C_ARG_DECL)
...
+                        // There is no known GObject library uses
+                        // hashtables of floats, so this is for values
+                        // only.
+                        gfloat *p = g_malloc(sizeof(double));
+                        *p = _keyval[j].v_float;
+                        _keyval[j].v_pointer = p;
+                    }
+                    else
+                        g_assert_not_reached();
+                }
```
Comment (LordYuuma) [inline src/gig_argument.c]: I'd abstract this part into a function, so that you can call it once with `key_type` and once with `value_type`.

**Comment by spk121** [comment]
This branch is stale and contains the old direct hash API which is no longer needed. I've pulled the relevant code into a new branch 31c90bac7b482c3789fdb2fef166c5e9b8497d9d .

**Pull request closed without merging**